test/src/test_dense.cc

```diff
@@ -0,0 +1,443 @@
+        test_mean_stddev<double>(seed, 203, 203, dn, expect_stddev);
+        test_mean_stddev<double>(seed, 203, 503, dn, expect_stddev);
+    }
+}
+
+
+class TestLSKGE3 : public ::testing::Test
+{
+    /*
+    Things that need to be tested:
+        1. Sketching the identity matrix works as expected.
+        2. Transposing the sketching operator works as expected.
+        3. Using a submatrix of the sketching operator.
+        4. sketching a submatrix of a data matrix.
+        5. row-major sketching
+            5.1 An exception should be raised if S0.layout != 
+                declared layout.
+
+    Things that don't need to be tested:
+        ?
+    */
+    protected:
+    
+    virtual void SetUp(){};
+
+    virtual void TearDown(){};
+
+    template <typename T>
+    static void sketch_eye(
+        uint32_t seed,
+        int64_t m,
+        int64_t d,
+        bool preallocate,
+        blas::Layout layout
+    ) {
+        // Define the distribution for S0.
+        RandBLAS::dense::DenseDist D = {
+            .family = RandBLAS::dense::DenseDistName::Gaussian,
+            .n_rows = d,
+            .n_cols = m
+        };
+
+        // Define the sketching operator struct, S0.
+        RandBLAS::dense::DenseSkOp<T> *S0_ptr;
+        std::vector<T> buff; // awkward.
+        if (preallocate) {
+            buff.resize(d * m, 0.0);
+            S0_ptr = new RandBLAS::dense::DenseSkOp<T>(
+                D, seed, 0, buff.data(), false, true, layout
+            );
+        } else {
+            S0_ptr = new RandBLAS::dense::DenseSkOp<T>(
+                D, seed, 0, buff.data(), false, true, layout
+            );
+        }
+
+        // define a matrix to be sketched, and create workspace for sketch.
+         bool is_colmajor = layout == blas::Layout::ColMajor;
+        std::vector<T> A(m * m, 0.0);
+        for (int i = 0; i < m; ++i)
+            A[i + m*i] = 1.0;
+        std::vector<T> B(d * m, 0.0);
+        int64_t lda = m;
+        int64_t ldb = (is_colmajor) ? d : m;
+
+        // Perform the sketch
+        RandBLAS::dense::lskge3<T>(
+            (*S0_ptr).layout,
+            blas::Op::NoTrans,
+            blas::Op::NoTrans,
+            d, m, m,
+            1.0, *S0_ptr, 0, 0, A.data(), lda,
+            0.0, B.data(), ldb
+        );
+
+        // check the result
+        RandBLAS_Testing::Util::buffs_approx_equal(B.data(), (*S0_ptr).buff, d*m);
+    }
+
+    template <typename T>
+    static void transpose_S(
+        uint32_t seed,
+        int64_t m,
+        int64_t d,
+        blas::Layout layout
+    ) {
+        // Define the distribution for S0.
+        RandBLAS::dense::DenseDist Dt = {
+            .family=RandBLAS::dense::DenseDistName::Gaussian,
+            .n_rows = m,
+            .n_cols = d
+        };
+        // Define the sketching operator struct, S0.
+        RandBLAS::dense::DenseSkOp<T> S0(Dt, seed, 0,
+            NULL, false, true, layout
+        );
```

Comment:
Default arguments are great. However, whether or not some default arguments are left to their default values depends heavily on context. The majority of users wouldn't care about setting ``layout``, since it defaults to ``ColMajor``. But if the user wants to specify "use a RowMajor layout, but only once it becomes necessary to allocate memory", then they have to provide the explicit defaults (here, ``NULL, false, true``, in order).

If we move ``layout`` earlier in the argument list then we run into the same problem with other arguments. For example, we usually have ``persistent=true``, but we might want ``persistent=false`` and to use column-major layout without explicitly stating as much. This is kind of a red herring though since I might be able to do away with ``layout`` in the ``DenseSkOp`` class in light of moving from "pointer offsets" to ``(i_op, j_op)``.

Here's a simpler example. At time of leaving this comment, the following function call works:
```
RandBLAS::sparse::SparseSkOp<T> sas(
    RandBLAS::sparse::SparseDistName::SASO,
    n_rows, n_cols, vec_nnz, key, ctr_offset
);
```
However, the following instantiation (of what seems to be an equivalent struct) fails
```
    RandBLAS::sparse::SparseSkOp<T> sas = {
        .family = RandBLAS::sparse::SparseDistName::SASO,
        .n_rows = n_rows,
        .n_cols = n_cols,
        .vec_nnz = vec_nnz,
        .key = key,
        .ctr_offset = ctr_offset
    };
```

Reply:
Hi. My suggestion is not bother too much with the order of the parameters. This may be difficult to maintain in a long term since your struct has lots of arguments. I you provide different kinds of constructors, which will have their own default arguments. Also, since you have a struct, you can always set values after the object is created:

```c++
RandBLAS::sparse::SparseSkOp<T> sas( ... );
sas.persistent = false;
```